<table width="100%" border="0" cellspacing="2" cellpadding="2">
 <tr align="left" valign="top">
  <td>
   <table align="left" border="0" cellspacing="0" cellpadding="0">
    <tr><td><font color="Green">Green</font>:&nbsp;</td><td>0 Frame Start Codon</td></tr>
    <tr><td><font color="Red">Red</font>:&nbsp;</td><td>0 Frame Stop Codon</td></tr>
    <tr><td><font color="Blue">Blue</font>:&nbsp;</td><td>Slippery Site</td></tr>
    <tr><td><font color="Orange">Orange</font>:&nbsp;</td><td>First -1 PTC</td></tr>
    <tr><td><font color="#FF0080">Maroon</font>:&nbsp;</td><td>SNP</td></tr>
   </table>
  </td>
 </tr>
 <tr align="left" valign="top">
 </tr>
</table>
% my $sp = $db->MySelect(statement => "SELECT species FROM gene_info WHERE accession = ?", vars => [$accession], type => 'single');
% print "TESTME: $sp<br>\n";
% my $in = $db->MySelect(statement => "SELECT orf_start, orf_stop, mrna_seq, direction, id FROM genome WHERE accession = ?", vars =>[$accession], type => 'row');
% my $mt = "mfe_$sp";
% $mt = "mfe_virus" if ($mt =~ /virus/);
% my $slips = $db->MySelect(statement =>"SELECT DISTINCT(start), bp_mstop FROM $mt WHERE accession = ? ORDER BY start", vars => [$accession],);
% my $snp_struct = {};
% if ($sp eq 'homo_sapiens') {
%   my $snp_data = $db->MySelect(statement =>"SELECT id, cluster_id, location, alleles FROM snp WHERE gene_acc = ?", vars => [$accession], type => 'list_of_hashes');
%  foreach my $snp_row (@{$snp_data}) {
%#      my $snp_start = $snp_row->{location} + $in->[0];
%      my $snp_start = $snp_row->{location};
%      my $snp_end = '';
%      if ($snp_start =~ /\.\./) {
%          ($snp_start, $snp_end) = split(/\.\./, $snp_start);
%          $snp_struct->{$snp_start} = $snp_end;
%          $snp_struct->{$snp_end} = $snp_row;  ## This is when a position evaluates to HASH
%      } else {
%          $snp_struct->{$snp_start} = $snp_row;
%      }
%  }
% }
% my (@slipsites, @minus);
% foreach my $s (@{$slips}) {
%   push (@slipsites, $s->[0]);
%   push (@minus, $s->[1]);
%  }
% my $str = HTMLMisc::Create_Pretty_mRNA(show_frame => $show_frame, accession => $accession, mrna_seq => $in->[2], orf_start => $in->[0], orf_stop => $in->[1], slipsites => \@slipsites, minus_bp => \@minus, snps => $snp_struct,);

<table bgcolor="#E8E8E8" align="center">
  <tr>
    <td>
    <strong>
      <pre>
% if ($show_frame ne 'none' and $show_frame ne 'all') {
Showing minus <% $show_frame %> reading frame(s).
% }
&gt;<% $accession %> | <% $genename %>
<% $str %>
      </pre>
    </strong>
    </font>
    </td>
    </tr>
    <tr><td>
      <form action="print_gene.html" enctype="multipart/form-data" method="POST">
      <input type="hidden" name="accession" value="<% $accession %>">
      <select name="show_frame"  onchange="this.form.submit()">
        <option value="none" <% (($show_frame eq 'none') ? 'SELECTED' : '') %>>none</option>
        <option value="zero" <% (($show_frame eq 'zero') ? 'SELECTED' : '') %>>zero frame</option>
        <option value="one" <% (($show_frame eq 'one') ? 'SELECTED' : '') %>>minus one frame</option>
        <option value="two" <% (($show_frame eq 'two') ? 'SELECTED' : '') %>>minus two frame</option>
        <option value="all" <% (($show_frame eq 'all') ? 'SELECTED' : '') %>>all</option>
      </select>
      </input>
      Show translated reading frame(s).
    </td></tr>
</form>
</table>
<p>The minimum free energy 'landscape' for <% $accession %>, black vertical lines delineate
the positions of slippery heptamers, the horizontal green line shows the mean Vienna RNAfold
value over the ORF, the horizontal red line shows the mean pknots value over the ORF,
the horzontal blue line shows the mean nupack value.  green, red, and blue lines delineate
the MFE over a sliding 105bp window.</p>
<& print_landscape.html, accession => "$accession", species => "$sp" &>
</form>   </tr>

% my $aux_info = $db->MySelect(statement => "SELECT publications, db_xrefs, gene_synonyms FROM gene_info WHERE accession = ?", vars => [$accession]);
% my $publications = $aux_info->[0]->[0];
% my @pubs = split(/\t/, $publications);
% my $xrefs = $aux_info->[0]->[1];
% my @xr = split(/\s+/, $xrefs);
% my $synon = $aux_info->[0]->[2];
% if (defined($synon)) {
<p>This gene is also known as: <% $synon %><br>
% }
% if (defined($xrefs)) {
This gene is referenced as:
<ul>
% foreach my $x (@xr) {
%  my ($name, $info) = split(/:/, $x);
%  if ($name =~ /GeneID/) {
   <a href="http://www.ncbi.nlm.nih.gov/gene?term=<% $info %>"><% $x %></a>
%  } elsif ($name =~ /HGNC/) {
   <a href="http://genenames.org/data/hgnc_data.php?hgnc_id=<% $info %>"><% $x %></a>
%  } elsif ($name =~ /HPRD/) {
   <a href="http://www.hprd.org/"><% $x %></a>
%  } elsif ($name =~ /MIM/) {
   <a href="http://www.ncbi.nlm.nih.gov/omim/<% $info %>"><% $x %></a>
%  } else {
  <% $x %>
%  }
% }
</p>
% }
% if (scalar(@pubs) > 0) {
<p>Below is a listing of publications associated with the refseq entry for this gene.</p>
<ul>
% foreach my $p (@pubs) {
%  my $q = $p;
%  $q =~ s/Reference: //g;
%  $q =~ s/ /\+/g;
  <li><a href="http://scholar.google.com/scholar?q=<% $q %>"><% $p %></a></li>
% }
</ul>
% }


<%init>
 my $statement = "SELECT species, genename FROM gene_info WHERE accession = ?";
 my $fun = $db->MySelect(statement => $statement, vars => [$accession,]);
 my $species = $fun->[0][0];
 my $genename = $fun->[0][1];
</%init>

<%args>
 $accession => "SGDID:S0004310"
 $translate => undef
 $show_frame => "none"
</%args>
