% if ($short or ($query and $search_species_limit)) {
<%perl>
my $query_statement;
my $entries;
if ($short) {
  $query_statement = qq"SELECT * FROM genome WHERE accession = ?";
  $entries = $db->MySelect(statement => $query_statement, type => 'hash', descriptor => 1, vars =>[$accession]);
} else {
  $query_statement = qq"SELECT * FROM genome WHERE ";
  if ($search_species_limit ne 'all') {
    $query_statement .= qq"species = '$search_species_limit' AND ";
#  } elsif ($search_species_limit ne 'all') {
#    $query_statement .= qq"species regexp '$search_species_limit' AND ";
  }
  my $mode = undef;
  if (defined($mode)) {   ## SOME SNP STUFF
    $query_statement .= qq"";
  } else {
    $query_statement .= qq"(species regexp '$query' OR genename regexp '$query' OR accession regexp '$query' OR locus regexp '$query' OR comment regexp '$query')";
  }
  $entries = $db->MySelect(statement => $query_statement, type => 'hash', descriptor => 1,);
}
my @entries_keys = ();
my $entries_count = 0;
foreach my $c (keys %{$entries}) {
  push(@entries_keys, $c);
  $entries_count++;
  my $slip_stmt = qq"SELECT count(distinct(start)), count(distinct(id)) FROM mfe WHERE accession = ?";
  my $slipsite_structure_count = $db->MySelect(statement => $slip_stmt, vars => [$entries->{$c}->{accession}], type => 'row',);
  $entries->{$c}->{slipsite_count} = $slipsite_structure_count->[0];
  $entries->{$c}->{structure_count} = $slipsite_structure_count->[1];
} ## End foreach
if ($entries_count == 0) {
  print "  <p>No entry was found.</p><br>\n";
}
my $species = $entries->{$entries_keys[0]}->{species};
my $accession = $entries->{$entries_keys[0]}->{accession};
my $long_species;
</%perl>
% if ($entries_count == 1) {
% $long_species = HTMLMisc::Make_Species($entries->{$entries_keys[0]}->{species});
% my $genename = $entries->{$entries_keys[0]}->{genename};
% my $accession = $entries->{$entries_keys[0]}->{accession};
% my $comment = $entries->{$entries_keys[0]}->{comment};
% my $omim_id = $entries->{$entries_keys[0]}->{omim_id};
% my $orf_start = $entries->{$entries_keys[0]}->{orf_start};
% my $orf_stop = $entries->{$entries_keys[0]}->{orf_stop};
<& gene_summary.html, accession => $accession, genename => $genename, accession => $accession, long_species => $long_species, species => $species, comment => $comment, omim_id => $omim_id &>
<& list_slipsites.html, accession => $accession, species => $species, comment => $comment, omim_id => $omim_id, orf_start => $orf_start, orf_stop => $orf_stop &>
<& print_gene.html, accession => $accession, translate => $translate &>
% } else { ### If there is not 0 nor 1 hit, then there are multiple.
<table width="100%" border="0" cellspacing="0" cellpadding="2">
  <tr bgcolor="#000000">
     <td></td>
    <td width="35%" nowrap>
      <div align="left"><strong><font color="#FFFFFF">Gene Name</font></strong></div>
    </td>
    <td nowrap width="10%">
      <div align="left"><strong><font color="#FFFFFF">Accession</font></strong></div>
      </td>
    <td nowrap width="15%">
      <div align="left"><strong><font color="#FFFFFF">Species</font></strong></div>
    </td>
    <td nowrap width="5%">
      <div align="center"><strong><font color="#FFFFFF">#Slips</font></strong></div>
    </td>
    <td nowrap width="35%">
      <div align="left"><strong><font color="#FFFFFF">Comments</font></strong></div>
    </td>
  </tr>
% foreach my $entry (sort keys %{$entries}) {
% my $species = $entries->{$entry}->{species};
% $long_species = HTMLMisc::Make_Species($entries->{$entry}->{species});
% $entries->{$entry}->{long_species} = $long_species;
% my $slip_stmt = qq"SELECT count(distinct(start)), count(distinct(id)) FROM mfe WHERE accession = ?";
% my $slipsite_structure_count = $db->MySelect(statement => $slip_stmt, vars => [$entries->{$entry}->{accession}], type => 'row');
% $entries->{$entry}->{slipsite_count} = $slipsite_structure_count->[0];
% }  ## End foreach my $entry
% foreach my $id (sort { $entries->{$b}->{slipsite_count} <=> $entries->{$a}->{slipsite_count} } keys %{$entries}) {
% my $genename = $entries->{$id}->{genename};
% my $accession = $entries->{$id}->{accession};
% my $comment = $entries->{$id}->{comment};
% my $omim_id = $entries->{$id}->{omim_id};
% my $slipsite_count = $entries->{$id}->{slipsite_count};
  <tr>
    <td valign="top">
	<a href="/search.html?short=1&accession=<% $accession %>&species=<% $species %>" rel="external" target="_blank"><img src="/html/external.png" border="0"></a>
  </td>
    <td valign="top" width="35%">
      <div align="left">
<a href="/search.html?short=1&accession=<% $accession %>&species=<% $species %>" rel="external" target="_blank"><% $genename %></a>
<& link_out.html, accession => $accession &>

  </div>
    </td>
    <td valign="top" width="10%">
      <div align="left">
	  <a href="/search.html?short=1&accession=<% $accession %>&species=<% $species %>" rel="external" target="_blank"><strong><% $accession %></strong></a>
      </div>
    </td>
    <td valign="top" width="15%">
      <div align="left"><i><% HTMLMisc::Make_Species($species) %></i></div>
    </td>
    <td valign="top" width="5%">
      <div align="center"><strong><% $slipsite_count %></strong></div>
    </td>
    <td valign="top" width="35%">
      <div align="left"><% $comment %></div>
    </td>
  </tr>
  <tr><td> </td></tr>

% } ## End foreach entry in entries
</table>
% } ## End the else there are multiple hits

% } else {  ## If you haven't filled out any of the form, then provide it.

<FORM action="search.html" enctype="multipart/form-data" method="POST">

<center>
  To begin browsing the PRFdb, please enter your search terms below.<br>
  <input type="text" name="query"  size="20">
  <& species_selector.html, select_name => "search_species_limit" &>
  <input type="submit" name="search submit" value="Search"></form>
</center>

A majority of the Accessions in the PRFdb come from the Mammalian Genome Collection.
It is available here: <a href="http://mgc.nci.nih.gov">http://mgc.nci.nih.gov</a>.
<hr>

<form method="post" action="search_blast.html" enctype="application/x-www-form-urlencoded">
Search the <select name="location">
 <option value="local">PRFdb</option>
 <option value="remote">NCBI</option>
</select> for an arbitrary sequence using BLAST.<br>
<textarea name="search_string" rows=12 cols=100 wrap="physical"></textarea>
<br>
<input type="submit" name="blastsearch" value="Perform Blast Search"></form></tr>


% }

<%init>
</%init>
<%args>
  $short => 0
  $accession => undef
  $query => undef
  $search_species_limit => undef
  $blastsearch => undef
  $translate => undef
</%args>
