% if ($query and $search_species_limit) {
<%perl>
my $query_statement = qq"SELECT * FROM genome WHERE ";
if (defined($search_species_limit)) {
  $query_statement .= qq"species = '$search_species_limit' AND ";
}
elsif ($search_species_limit ne 'all') {
  $query_statement .= qq"species regexp '$search_species_limit' AND ";
}
my $mode = undef;
if (defined($mode)) {   ## SOME SNP STUFF
  $query_statement .= qq"";
}
else {
  $query_statement .= qq"(species regexp '$query' OR genename regexp '$query' OR accession regexp '$query' OR locus regexp '$query' OR comment regexp '$query')";
}
my $entries = $db->MySelect({statement => $query_statement, type => 'hash', descriptor => 1,});
my @entries_keys = ();
foreach my $c (keys %{$entries}) {
  push(@entries_keys, $c);
  my $slip_stmt = qq"SELECT count(distinct(start)), count(distinct(id)) FROM mfe WHERE accession = ?";
  my $slipsite_structure_count = $db->MySelect({statement => $slip_stmt, vars => [$entries->{$c}->{accession}], type => 'row',});
  	$entries->{$c}->{slipsite_count} = $slipsite_structure_count->[0];
	$entries->{$c}->{structure_count} = $slipsite_structure_count->[1];
    } ## End foreach
    my $entries_count = scalar keys %{$entries};
    if ($entries_count == 0) {
       print "  <p>No entry was found.</p><br>\n";
#    } elsif ($entries_count == 1) {  
#       print "   <p>1 match was found.</p>\n";
#    } else {
#       print "   <p><% $entries_count %> matches were found.</p><br>\n";
    }
my $long_species;
</%perl>
<& list_slipsites.html, accession => $accession, species => $species &>
% if ($entries_count == 1) {
% $long_species = $entries->{$entries_keys[0]}->{species};
% $long_species =~ s/_/ /g;
% $long_species = ucfirst($long_species);
<table width="100%" border="0" cellspacing="2" cellpadding="2">
 <tr align="left" valign="top">
   <td width="50%">
    <strong>
     <a href="/help_genome#help_genename" title="The gene name is provided by genbank or the SGD" rel="external" target="_blank">Gene Name:</a>
    </strong> <% $entries->{$entries_keys[0]}->{genename} %>
   <br>
   <strong>
    <a href="/help_detail#help_accession" title="Check the help for accession" rel="external" target="_blank">Accession:</a>
   </strong> <% $entries->{$entries_keys[0]}->{accession} %><br>
   <strong>
    <a href="/help_detail#help_species" title="Check the help for species" rel="external" target="_blank">Species:</a>
   </strong><i> <% $long_species %></i><br>
   <strong>
    <a href="/help_genome#help_comments" title="Check the help for the comment field" rel="external" target="_blank">Comments:</a>
   </strong> <% $entries->{$entries_keys[0]}->{comment} %><br>
  </td>
  <td width="50%">
% my ($short_accession, $accession);
% if ($entries->{$entries_keys[0]}->{accession} =~ /^SGDID/) {
%  $accession = $entries->{$entries_keys[0]}->{accession};
%  $short_accession = $accession;
%  $short_accession =~ s/^SGDID\://g;
   <a href="http://db.yeastgenome.org/cgi-bin/locus.pl?locus=<% $short_accession %>" title="View <% $accession %> at yeastgenome.org" rel="external" target="_blank">SGD Link</a>
% } elsif ($accession =~ /^BC/) {
   <a href="http://www.ncbi.nlm.nih.gov/sites/entrez?db=nuccore&cmd=&term=<% $accession %>&go=Go" title="View <% $accession %> at NCBI" rel="external" target="_blank">GenBank Link</a>
% }
% if (defined($entries->{$entries_keys[0]}->{omim_id})) {
   , <a href="http://www.ncbi.nlm.nih.gov/entrez/dispomim.cgi?id=<% $entries->{$entries_keys[0]}->{omim_id} %>" title="View associated OMIM:<% $entries->{$entries_keys[0]}->{omim_id} %> at NCBI" rel="external" target="_blank">OMIM Link</a>
% }
   <br>
   Run <a href="/search_local_blast?accession=<% $accession %>" title="Run BLAST to search the PRFdb for sequences similar to this gene.">Local Blast</a>
 or
   <a href="/remote_blast?accession=<% $accession %>" title="Run BLAST to search the NR database for sequences similar to this gene.">Remote Blast</a> using this gene.</br>
  <br>
% my $table = "landscape_$entries->{$entries_keys[0]}->{species}";
% my $landscape = $db->MySelect("SELECT count(id) FROM $table WHERE accession = '$accession'");
% if ($landscape->[0][0] > 0) {
  Graph </strong><a href="/landscape?accession=<% $accession %>" title="View a graph of putative Minimum Free Energy with respect to position." rel="external" target="_blank">MFE minima</a> or
% }
   <a href="/download_sequence?accession=<% $accession %>" title="Download a fasta file containing this gene.">Download Sequence</a><br>
  </td>
 </tr>
</table>
<& print_gene.html, accession => $accession &>

% }


% } elsif ($blastsearch) {



% } else {



<FORM action="search.html" enctype="multipart/form-data" method="POST">

<center>
  To begin browsing the PRFdb, please enter your search terms below.<br>
  <input type="text" name="query"  size="20"> 
  <& species_selector, select_name => "search_species_limit" &>
  <input type="submit" name="search submit" value="Search"></form>
</center>

A majority of the Accessions in the PRFdb come from the Mammalian Genome Collection.
It is available here: <a href="http://mgc.nci.nih.gov">http://mgc.nci.nih.gov</a>.
<hr>

Search the PRFdb for an arbitrary sequence using BLAST.<br>
<form method="post" action="search_blast.html" enctype="application/x-www-form-urlencoded">
<textarea name='blastsearch' rows=12 cols=100 wrap='physical'></textarea>
<br>
<input type="submit" name="blastsearch" value="Perform Blast Search"></form></tr>


% }

<%init>
</%init>
<%args>
  $query => undef
  $search_species_limit => 'saccharomyces_cerevisiae'
  $blastsearch => undef
</%args>
