 </td>
 <td width="70%">
 <p>
  Below is an example detailed output one can receiving by doing a
  search for 'EST2' and following the link to the putative PRF signal
  at position 1995. Click on any portion of the page below to receive
  a short description of that field.
 </p>

    <a name="example">Example entry from the PRFdb.</a><br>
    <table width="100%" border="0" align="center" cellpadding="2"
  cellspacing="1">

      <tr valign="top" bgcolor="#E9EBF7">
	<td width="20%">

	    <div align="right"><a href="#species"><strong>Species: </strong></a></div>
	    </td>
	<td>
	  <i>
	    Saccharomyces cerevisiae</i>
	  </td>
	<td width="40%" rowspan="12" valign="top">
	    <div align="left"><a href="#graph"><img src="/html/SGDID:S0004310-1995-443821.png" border="2"></a></div>
	    </td>
      </tr>

      <tr valign="top" bgcolor="#E9EBF7">
	<td width="20%">
	    <a href="#accession"><div align="right"><strong>Accession: </strong></div></a>
	    </td>
	<td>
	    <a href="/browse?accession=SGDID:S0004310">SGDID:S0004310</a>
	    </td>
      </tr>

      <tr valign="top" bgcolor="#E9EBF7">
	<td width="20%">
	  <a href="#position"><div align="right"><strong>Position: </strong></div></a>
	    </td>
	<td>
	  1995</td>

      </tr>

      <tr valign="top" bgcolor="#E9EBF7">
	<td width="20%">
	  <a href="#algorithm"><div align="right"><strong>Algorithm: </strong></div></a>
	    </td>
	<td>
	  pknots</td>

      </tr>

      <tr valign="top" bgcolor="#E9EBF7">
	<td width="20%">
	  <a href="#bacode"><div align="right"><strong>Barcode: </strong></div><a/>
	    </td>
	<td>
	  112233</td>

      </tr>

      <tr valign="top" bgcolor="#E9EBF7">
	<td width="20%">
	  <a href="#seqlength"><div align="right"><strong>Sequence Length: </strong></div></a>
	    </td>
	<td>
	  100</td>

      </tr>

      <tr valign="top" bgcolor="#E9EBF7">
	<td width="20%">
	  <a href="#slipsite"><div align="right"><strong>Slippery Site: </strong></div></a>
	    </td>
	<td>
	  UUUAAAA</td>

      </tr>


      <tr valign="top" bgcolor="#E9EBF7">
	<td width="20%">
	  <a href="#bp"><div align="right"><strong>Base Pairs: </strong></div></a>
	    </td>
	<td>
	  21</td>

      </tr>

      <tr valign="top" bgcolor="#E9EBF7">
	<td width="20%">
	  <a href="#mfe"><div align="right"><strong>MFE: </strong></div></a>
	    </td>
	<td>
	    -16.3 kcal/mol
	    </td>

      </tr>

      <tr valign="top" bgcolor="#E9EBF7">
	<td width="20%">
	  <a href="#z"><div align="right"><strong>Z Score: </strong></div></a>
	    </td>
	<td>
	  -0.73  </td>

      </tr>

      <tr valign="top" bgcolor="#E9EBF7">
	<td width="20%">
	  <a href="#rand_mfe"><div align="right"><strong>Randomized Mean MFE: </strong></div></a>
	    </td>
	<td>
	    -13.9 &plusmn; 0.3 kcal/mol
	    </td>

      </tr>

      <tr valign="top" bgcolor="#E9EBF7">
	<td width="20%">
	  <a href="#ppcc"><div align="right"><strong>PPCC: </strong></div></a>
	    </td>
	<td>
	  0.9632</td>

      </tr>
</table>


    <table align="center" width="100%" border="0" cellpadding="2" cellspacing="0" bgcolor="#E9EBF7">

      <tr>
	<td width="10%"></td>
	<td>
	    <font face="Courier New, Courier, mono"
	    size="-1">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;10&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;20&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;30&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;40&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;50&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;60&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;70&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;80&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;90&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;100</font>

        </td>
      </tr>

      <tr>
        <td width="10%"><a href="#seq">Sequence</a></td>
	<td>
    <font face="Courier New, Courier, mono"
    size="-1">C<font color="blue">U</font><font color="blue">G</font><font color="blue">G</font><font color="blue">C</font><font color="blue">U</font><font color="blue">G</font>ACGAUUUCCUUAUAAUAUCAA<font color="blue">C</font><font color="blue">A</font><font color="blue">G</font>A<font color="blue">C</font><font color="blue">C</font><font color="blue">A</font>A<font color="red">C</font><font color="red">A</font><font color="red">G</font><font color="red">C</font><font color="red">A</font><font color="red">A</font><font color="red">G</font><font color="red">U</font>GAUCAAUAUCAAAAA<font color="red">G</font><font color="red">C</font><font color="red">U</font><font color="red">U</font><font color="red">G</font><font color="red">C</font>CA<font color="red">U</font><font color="red">G</font><font color="green">G</font><font color="green">G</font><font color="green">C</font>GGA<font color="green">U</font><font color="green">U</font><font color="green">U</font><font color="green">C</font>AAAAAUAUAAUGC<font color="green">G</font><font color="green">A</font><font color="green">A</font><font color="green">A</font><font color="green">G</font><font color="green">C</font><font color="green">C</font>A</font>

      <font face="Courier New, Courier, mono" size="-3">
	    <a href="/download_subseq?mfeid=443840">Download
	    subsequence</a>
	      </font>
        </td>
      </tr>

      <tr>
        <td width="10%"><a href="#parens">Parens</a></td>
	<td>
	    <font face="Courier New, Courier, mono"
	    size="-1">.<font color="blue">(</font><font color="blue">(</font><font color="blue">(</font><font color="blue">(</font><font color="blue">(</font><font color="blue">(</font>.....................<font color="blue">)</font><font color="blue">)</font><font color="blue">)</font>.<font color="blue">)</font><font color="blue">)</font><font color="blue">)</font>.<font color="red">(</font><font color="red">(</font><font color="red">(</font><font color="red">(</font><font color="red">(</font><font color="red">(</font><font color="red">(</font><font color="red">(</font>...............<font color="red">)</font><font color="red">)</font><font color="red">)</font><font color="red">)</font><font color="red">)</font><font color="red">)</font>..<font color="red">)</font><font color="red">)</font><font color="green">(</font><font color="green">(</font><font color="green">(</font>...<font color="green">(</font><font color="green">(</font><font color="green">(</font><font color="green">(</font>.............<font color="green">)</font><font color="green">)</font><font color="green">)</font><font color="green">)</font><font color="green">)</font><font color="green">)</font><font color="green">)</font>.</font>

	      <font face="Courier New, Courier, mono" size="-3">
		    <a href="/download_parens?mfeid=443840">Download
		    brackets</a>
		      </font>
	      </td>
      </tr>

      <tr>
        <td width="10%"><a href="#parsed">Parsed</a></td>
       <td>
	   <font face="Courier New, Courier, mono"
	   size="-1">.<font color="blue">1</font><font color="blue">1</font><font color="blue">1</font><font color="blue">1</font><font color="blue">1</font><font color="blue">1</font>.....................<font color="blue">1</font><font color="blue">1</font><font color="blue">1</font>.<font color="blue">1</font><font color="blue">1</font><font color="blue">1</font>.<font color="red">2</font><font color="red">2</font><font color="red">2</font><font color="red">2</font><font color="red">2</font><font color="red">2</font><font color="red">2</font><font color="red">2</font>...............<font color="red">2</font><font color="red">2</font><font color="red">2</font><font color="red">2</font><font color="red">2</font><font color="red">2</font>..<font color="red">2</font><font color="red">2</font><font color="green">3</font><font color="green">3</font><font color="green">3</font>...<font color="green">3</font><font color="green">3</font><font color="green">3</font><font color="green">3</font>.............<font color="green">3</font><font color="green">3</font><font color="green">3</font><font color="green">3</font><font color="green">3</font><font color="green">3</font><font color="green">3</font>.</font>

	     <font face="Courier New, Courier, mono" size="-3">
	           <a href="/download_parsed?mfeid=443840">Download
	           parsed</a>
		     </font>
	     </td>
      </tr>

      <tr>
	<td width="10%"><a href="#feynman">Feynman</a></td>
        <td>

          <embed src="/images/feynman/00/04/SGDID:S0004310-443840.svg"
          width = 821 height = 98.75 border = "0">

	      <font face="Courier New, Courier, mono" size="-3">
		    <a href="/download_bpseq?mfeid=443840">Download
		    bpseq</a>
		      </font>
        </td>
      </tr>

    </table>

 <p>
  Most information is kept on the 'detail' page.  Once a particular
  gene and putative PRF site has been chosen, it is possible to view
  some more specific information about that PRF signal here.<br>
  Each group of information concerns one folding performed for the
  given PRF signal.  Thus we see once again the gene's name, species,
  accession, and position.<br>
  The 'algorithm' field specifies which minimum free energy folding
  program was used to fold this specific sequence window.<br>
  'Barcode' provides a shorthand notation to quickly identify how many
  stems exist in the putative secondary structure and how they are
  oriented.  In this notation, a simple stem is '11' and a H-type
  pseudoknot is '1212'.<br>
  'Sequence length' and 'Slippery site' provide some information about
  the context of the given sequence window.<br>
  'Base pairs' and 'MFE' show the measurements provided by the mimimum
  free energy heuristics of the given prediction algorithm.<br>
  'Z score,' 'Randomized Mean MFE,' and 'PPCC' provide some
  statistical information about how this particular sequence window
  compares to a distribution of randomized sequences folded in a
  similar fashion.<br>
  'Z score' is defined as (Actual mfe - the mean mfe of the randomized
  sequences) / the standard deviation of the randomized sequences.
  As a result, the more negative a given Z score is, the more
  interesting it should prove.<br>
  PPCC is a measurement which can give a sense of how likely the Z
  score is to be useful.  A Z score assumes its data is normally
  distributed.  The PPCC provides a measurement of how close to
  'normal' a given set of values actually is.<br>
  These last few fields are summarized in the graph on the right side
  of the table.  This bar graph shows a series of 'bins' and what
  percentage of the randomized sequences have minimum free energies
  falling into each 'bin.'  A red line provides a guide to how many
  entries one would expect in each bin if the minimum free energy
  values were normally distributed.  The green line shows where on
  this distribution the actual minimum free energy lies.  If the
  actual MFE is more negative than the mean of the distribution of
  random values, that is a good sign for the significance of the
  putative PRF signal.<br>
  Below this graph lies a sequence view, and a few ways to visualize
  the putative mRNA secondary structure, including a bracket notation,
  a numeric stem diagram, and a linear feynman image.<br>
<p><hr></p>

 <p>
 <a name="species">Species</a><br>
   The 'species' field tells just that.  Currently the only likely
   species one will see are <i>Saccharomyces cerevisiae</i>, <i>Homo
   sapiens</i>, and <i>Mus musculus</i>
 </p>

 <p>
 <a name="graph">Graph</a><br>
  Every sequence window of the PRFdb is randomized using one of a few
  potential randomization strategies (shuffling, maintain nucleotide
  frequencies, maintain dinucleotide frequencies, maintain reading
  frame).  In each instance, the given sequence window is randomized
  and refolded using the given algorithm 100 times.  This plot shows
  how often the resulting randomized sequences result in each provided
  minimum free energy.  Thus, we can see that 10 randomized sequences
  of the 100 have a minimum free energy of 15.4 Kcal/mol when folded
  with pknots.  The red line provides an idealized normal distribution
  of how many sequences one would expect of each range if the
  sequences are properly randomized.  Finally, the green bar shows the
  actual minimum free energy calculated using pknots at this
  position.  The further to the left of the mean this value reaches,
  the more significant it should be with respect to its randomized
  sequences.
 </p>

 <p>
 <a name="accession">Accession</a><br>
  The accession provides the genbank or SGD accession of the given
  gene.  Clicking on the accession itself will take you back to the
  gene view.
 </p>

 <p>
 <a name="position">Position</a><br>
  This number defines how many bases from the beginning of the mRNA
  the actual PRF signal may be found.  In order to get a sense of
  where this is in the gene, it may be most useful to follow the
  accession link and look at the MFE minima graph or the text
  representation of the message.
 </p>

 <p>
 <a name="algorithm">Algorithm</a><br>
  The algorithm may be one of 'pknots,' 'nupack,' or 'hotknots.'
  These secondary structure prediction programs provide a putative
  mRNA structure as well as a putative minimum free energy.
 </p>

 <p>
 <a name="barcode">Barcode</a><br>
   The 'barcode' is a shorthand representation of the secondary
   structureo of a given sequence window.  A sequence consisting of a
   single stem would thus show up as '11' while an H-type pseudoknot
   is '121' and a SARS-like pseudoknot with a secondary stem would be
   '121332'.
 </p>

 <p>
 <a name="seqlength">Sequence Length</a><br>
  The sequence lenght is usally 100 bases.  However this is
  variable in <i>Saccharomyces cerevisiae</i> sequences.
 <p>


 <p>
 <a name="slipsite">Slipsite</a><br>
  There are 24 separate slippery heptamers which follow the motif
  NNNWWWH.  This field shows which heptamer is used by this sequence
  window.
 </p>

 <p>
 <a name="bp">Number of Base Pairs</a><br>
  This field provides the number of bases which are paired in the
  sequence.
 </p>


 <p>
 <a name="mfe">MFE</a><br>
  The minimum free energy is calculated by summing the free energy
  change associated with a group of known base pair configurations
  including: stacking bases, GC/AU/GU base pairs, internal bulges,
  hairpin loops, and known motifs like tetraloops.  Each secondary
  structure prediction algorithm makes different assumptions about
  these.<br>
  The putative minimum free energy calculated by pknots for this
  sequence window is provided here.
 </p>

 <p>
 <a name="z">z score</a><br>
  The z score of a given sequence is calcuated as the difference
  between the mean mfe of the randomized sequences and the mfe of
  the given sequence divided by the standard deviation of the
  randomized mfes.  Thus values which are more negative are more
  significant with respect to the randomized sequences.
 </p>

 <p>
 <a name="rand_mfe">Randomized MFE</a><br>
   This field shows the mean of the randomized mfe values.  The
   standard error is reported.  It is calculated as the standard
   deviation divided by the square root of the number of iterations.
   In most situations this is the standard deviation divided by 10.
   As a result this is a nice summary of what is going into the z
   score calculation.
 </p>

 <p>
 <a name="ppcc">PPCC</a><br>
  PPCC stands for Probability Plot Correlation Coefficient.  This is
  calculated by taking the Correlation coefficient of an idealized
  normal distribution and the actual distribution provided by the
  randomized mfe values.  Each value of the idealized normal
  distribution is i/n+1 where both i and n iterate from 1 until
  (usually) 100.  Each value of the actual distribution is calculated
  by sorting the randomized mfe values and calculating the upper
  probability of a normal distribution of the mfe value - the mean mfe
  divided by the standard deviation.  As the given coefficient
  approaches 1.0, the more similar that array of randomized values is
  to a normal distribution.
 </p>

 <p>
 <a name="seq">Sequence Window</a>
  The actual sequence window folded.  You can download a fasta file of
  it to the right.
 </p>

 <p>
 <a name="parens">Parens</a>
  Some MFE prediction algorithms present a parenthesis output
  shorthand where '()' describes bases participating in stems and '{}'
  describes bases participating in stems which are a part of a
  pseudoknot.  For those who prefer this output, this is provided, as
  well as a link to download this output.
 </p>

 <p>
 <a name="parsed">Parsed</a>
  The 'parsed' output is a shorthand way to describe each stem.  Each
  base is either a '.' or number of a stem.
 </p>

 <p>
 <a name="feynman">Feynman</a>
  A feynman diagram of this secondary structure is generated for each
  sequence window.  It is possible to download the .bpseq
  format of the secondary structure beside this image.
 </p>

 </td>
 <td width="15%"></td>
</tr>
</table>
